tests/ping.go

```diff
@@ -0,0 +1,103 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package tests
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	expect "github.com/google/goexpect"
+	"google.golang.org/grpc/codes"
+	"k8s.io/utils/net"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+)
+
+const promptExpression = `(\$ |\# )`
+const termNLExpression = "\r\n"
+
+var (
+	shellSuccess = regexp.MustCompile("\n0" + termNLExpression + ".*" + promptExpression)
+	shellFail    = regexp.MustCompile("\n[1-9].*" + termNLExpression + ".*" + promptExpression)
+)
+
+// PingFromVMConsole performs a ping through the provided VMI console.
+// Optional arguments for the ping command may be provided, overwirting the default ones.
+// (default ping options: "-c 1, -w 5")
+// Note: The maximum overall command timeout is 10 seconds.
+func PingFromVMConsole(vmi *v1.VirtualMachineInstance, ipAddr string, args ...string) error {
+	const maxCommandTimeout = 10
+
+	pingString := "ping"
+	if net.IsIPv6String(ipAddr) {
+		pingString = "ping -6"
+	}
+
+	if len(args) == 0 {
+		args = []string{"-c 1", "-w 5"}
+	}
+	args = append([]string{pingString, ipAddr}, args...)
+	cmdCheck := strings.Join(args, " ") + "\n"
+
+	err := vmiConsoleExpectBatch(vmi, []expect.Batcher{
+		&expect.BSnd{S: "\n"},
+		&expect.BExp{R: promptExpression},
+		&expect.BSnd{S: cmdCheck},
+		&expect.BExp{R: promptExpression},
+		&expect.BSnd{S: "echo $?\n"},
+		&expect.BCas{C: []expect.Caser{
+			&expect.Case{
+				R: shellSuccess,
+				T: expect.OK(),
+			},
+			&expect.Case{
+				R: shellFail,
+				T: expect.Fail(expect.NewStatus(codes.Unavailable, "ping failed")),
+			},
+		}},
+	}, maxCommandTimeout)
+	if err != nil {
+		return fmt.Errorf("Failed to ping VMI %s, error: %v", vmi.Name, err)
+	}
+	return nil
+}
+
+func vmiConsoleExpectBatch(vmi *v1.VirtualMachineInstance, expected []expect.Batcher, wait int) error {
```

Comment:
What is the difference between this method and `test.CheckForTextExpecter`?

Reply:
I explained it in the commit message. This one is not using the “safe” in order to work with the caser.

Reply:
You caught me, I didn't read the commit messge. Thanks for explaining again. I think it is better to somehow reuse the code of `test.CheckForTextExpecter` instead of copy paste it. But up to you.